sql: fix drop database - sequence ownership bug

arulajmani opened the pull request:
Previously, `DROP DATABASE CASCADE` would not work if the database
contained a sequence owned by a table in the database. This would
happen because of two separate reasons:
- If the sequence was dropped before the table, the table would try to
"double drop" the sequence as it owned it; this would result in an
error.
- If the table was dropped before the sequence, the sequence would try
to remove the ownership dependency from the table descriptor, which had
already been dropped; this would also result in an error.

This PR addresses both these issues separately. Sequences are no longer
double dropped when dropping tables. Additionally, no attempt is made
to remove the ownership dependency from the table descriptor if the
table descriptor has already been dropped.

Fixes #50712

Release note (bug fix): `DROP DATABASE CASCADE` now works as expected
even when the database has a sequence with an owner in it.

arulajmani committed:
sql: fix drop database - sequence ownership bug

Previously, `DROP DATABASE CASCADE` would not work if the database
contained a sequence owned by a table in the database. This would
happen because of two separate reasons:
- If the sequence was dropped before the table, the table would try to
"double drop" the sequence as it owned it; this would result in an
error.
- If the table was dropped before the sequence, the sequence would try
to remove the ownership dependency from the table descriptor, which had
already been dropped; this would also result in an error.

This PR addresses both these issues separately. Sequences are no longer
double dropped when dropping tables. Additionally, no attempt is made
to remove the ownership dependency from the table descriptor if the
table descriptor has already been dropped.

Fixes #50712

Release note (bug fix): `DROP DATABASE CASCADE` now works as expected
even when the database has a sequence with an owner in it.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/50744)
<!-- Reviewable:end -->

arulajmani commented:
For some reason, the tests I added aren't moving past the `DROP DATABASE CASCADE` line on 3node-tennant; I've added a blocklist on the whole file for now, not sure if this is okay/why the test isn't working.

rohany commented:
Note, use the `cockroachdb/errors` package (`errors.Newf`) rather than `fmt.Errorf`

lucy-zhang commented:
> For some reason, the tests I added aren't moving past the DROP DATABASE CASCADE line on 3node-tennant; I've added a blocklist on the whole file for now, not sure if this is okay/why the test isn't working.

I haven't looked closely at these changes yet, but this sounds potentially bad. Would you mind digging further into what's getting stuck? I wonder if it's related to the jobs.

solongordon reviewed (changes requested):
Agree that we should dig into what's going wrong with that test config.

The changes look reasonable to me though I don't have a deep understanding of the `queueJob` parameter. It would be great for @lucy-zhang to sign off on that change, i.e. that `queueJob` is now being set to false when dropping sequences during a DROP DATABASE.

Could you also add analogous tests for DROP DATABASE without CASCADE? Looks like this is not covered right now.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/50744#-:-MB0VjH36uuDT0uDAxKQ:b-ejryru)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @solongordon)



<!-- Sent from Reviewable.io -->

arulajmani commented:
Couldn't find anything that immediately jumped out w.r.t the 3node-tenant config from the stack traces. Looks like the problem is not related to sequences -- it fails on a db containing a single, regular table as well. I've opened https://github.com/cockroachdb/cockroach/issues/50840

arulajmani reviewed (commented):
Added a test for `DROP DATABASE` without CASCADE as well. Should be RFAL. 

Note on the `queueJob` -- as part of my change, I plumbed it down so that dropping a sequence owned by a table respects the `queueJob` of the table that owned it. It doesn't have an affect on the bug. 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/50744#-:-MB5xBLkBEs1UD7P1Qa8:b-t7217s)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @lucy-zhang)



<!-- Sent from Reviewable.io -->

solongordon reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 4 of 5 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/50744#-:-MBFsot3-DygxxkwexVE:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @lucy-zhang and @solongordon)



<!-- Sent from Reviewable.io -->

arulajmani commented:
bors r=solongordon
